research/conditional/moe_layers/expert_choice.py

```diff
@@ -363,42 +151,47 @@ def extract_with_linear_bmm(
                 seq_len, batch_size, n_exp * topk
             )
         with measure_time(self, "shuffle_preprocess"):
-            x = torch.bmm(x, one_hot_perm).reshape(seq_len, dmodel, n_exp, topk)
+            x = torch.bmm(x, one_hot_perm).reshape(seq_len, self.dmodel, n_exp, topk)
 
-        # seq_len dmodel n_exp topk, n_exp dmodel exp_size
-        # x * weight (BROAD n_exp, MUL=dmodel, N=seq_len, M=exp_size)
-        # -> n_exp seq_len topk exp_size
         with measure_time(self, "lin1_perm"):
-            x = x.permute(2, 0, 3, 1).reshape(n_exp, seq_len * topk, dmodel)
-        with measure_time(self, "lin1"):
-            x = torch.bmm(x, weight)
-
-        assert x.shape == (n_exp, seq_len * topk, exp_size)
+            x = x.permute(2, 0, 3, 1).reshape(n_exp, seq_len * topk, self.dmodel)
         return x, one_hot_perm
 
-    def gating_postprocess_onehot_with_linear(self, x, topk_values, one_hot, weight):
-        with measure_time(self, "gating_postprocess_with_linear"):
-            return einsum(
-                "n_exp topk seq_len exp_size, n_exp topk seq_len, "
-                "n_exp topk seq_len batch_size, n_exp exp_size doutput"
-                "-> batch_size seq_len doutput",
-                x,
-                topk_values,
-                one_hot,
-                weight,
-            )
+    def extract_index_select(self, x: torch.Tensor, topk, topk_indices: torch.Tensor):
+        batch_size, _, _ = x.shape
+        # flatten x s. t. first dimension is tokens instead of batch_size x seq_len
+        with measure_time(self, "first_flatten"):
+            x = x.flatten(start_dim=0, end_dim=1)
+        with measure_time(self, "index_select"):
+            x = torch.index_select(x, dim=0, index=topk_indices.flatten())
+        with measure_time(self, "reshape"):
+            x = x.reshape((self.n_experts, topk, self.dmodel))
+        return x, None
+
+    # postprocess implementations
 
-    def gating_postprocess_bmm(self, x, topk_values, one_hot, weight):
-        n_exp, exp_size, dmodel = weight.shape
-        seq_len, batch_size, _ = one_hot.shape
+    @time_measured("postprocess_einsum")
+    def merge_einsum(
+        self, x, batch_size, topk, seq_len, topk_values, topk_indices, one_hot
```

Comment:
jesteś w stanie `batch_size, topk, seq_len` wziąć z `one_hot`

Reply:
to samo